module/resources.go

```diff
@@ -0,0 +1,547 @@
+	}
+	if resInfo.Constructor == nil {
+		return fmt.Errorf("invariant: no constructor for %q", conf.Model)
+	}
+
+	res, err := resInfo.Constructor(ctx, deps, *conf, resLogger)
+	if err != nil {
+		return err
+	}
+
+	// If context has errored, even if construction succeeded we should close the resource and
+	// return the context error.  Use shutdownCtx because otherwise any Close operations that rely
+	// on the context will immediately fail.  The deadline associated with the context passed in to
+	// this function is rutils.GetResourceConfigurationTimeout, which is propagated to AddResource
+	// through gRPC.
+	if ctx.Err() != nil {
+		m.logger.CDebugw(ctx, "resource successfully constructed but context is done, closing constructed resource",
+			"err", ctx.Err().Error())
+		return multierr.Combine(ctx.Err(), res.Close(m.shutdownCtx))
+	}
+
+	var passthroughSource rtppassthrough.Source
+	if p, ok := res.(rtppassthrough.Source); ok {
+		passthroughSource = p
+	}
+
+	m.registerMu.Lock()
+	defer m.registerMu.Unlock()
+
+	coll, ok := m.collections[conf.API]
+	if !ok {
+		return fmt.Errorf("module cannot service api: %s", conf.API)
+	}
+
+	// If adding the resource name to the collection fails, close the resource and return an error.
+	if err := coll.Add(conf.ResourceName(), res); err != nil {
+		return multierr.Combine(err, res.Close(ctx))
+	}
+
+	m.logger.Infof("DBG. Added res to loggers: %p %v", res, res.Name())
+	m.resLoggers[res] = resLogger
+
+	// add the video stream resources upon creation
+	if passthroughSource != nil {
+		m.streamSourceByName[res.Name()] = passthroughSource
+	}
+
+	for _, dep := range deps {
+		// If the dependency is in the `resLogger` it is a "local resource". And we must track
+		// reconfigures on our dependencies as that invalidates resource handles.
+		if _, exists := m.resLoggers[dep]; exists {
+			m.internalDeps[dep] = append(m.internalDeps[dep], resConfigureArgs{
+				toReconfig: res,
+				conf:       conf,
+				depStrings: depStrings,
+			})
+		}
+	}
+
+	return nil
+}
+
+func (m *Module) removeResource(ctx context.Context, resName resource.Name) error {
+	slowWatcher, slowWatcherCancel := utils.SlowGoroutineWatcher(
+		30*time.Second, fmt.Sprintf("module resource %q is taking a while to remove", resName), m.logger)
+	defer func() {
+		slowWatcherCancel()
+		<-slowWatcher
+	}()
+
+	m.registerMu.Lock()
+	coll, ok := m.collections[resName.API]
+	if !ok {
+		m.registerMu.Unlock()
+		return fmt.Errorf("no grpc service for %+v", resName)
+	}
+
+	res, err := coll.Resource(resName.Name)
+	if err != nil {
+		m.registerMu.Unlock()
+		return err
+	}
+	m.registerMu.Unlock()
+
+	if err := res.Close(ctx); err != nil {
+		m.logger.Error(err)
+	}
+
+	m.registerMu.Lock()
+	defer m.registerMu.Unlock()
+	delete(m.streamSourceByName, res.Name())
+	delete(m.activeResourceStreams, res.Name())
+
+	// Dan: Log if there are still resources that depend on us? We assume the viam-server forbids
```

Comment:
viam-server does close down dependents first, so anything still existing would be unexpected